pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go

```diff
@@ -0,0 +1,396 @@
+			}
+			if len(waypoints) > 0 {
+				wp := waypoints[0]
+				w.Waypoint = &workloadapi.GatewayAddress{
+					Destination: &workloadapi.GatewayAddress_Address{
+						Address: a.toNetworkAddress(wp.Addresses[0].String()),
+					},
+					// TODO: look up the HBONE port instead of hardcoding it
+					HboneMtlsPort: 15008,
+				}
+			}
+
+			if td := spiffe.GetTrustDomain(); td != "cluster.local" {
+				w.TrustDomain = td
+			}
+			w.WorkloadName, w.WorkloadType = se.Name, workloadapi.WorkloadType_POD // XXX(shashankram): HACK to impersonate pod
+			w.CanonicalName, w.CanonicalRevision = kubelabels.CanonicalService(se.Labels, w.WorkloadName)
+
+			if se.Annotations[constants.AmbientRedirection] == constants.AmbientRedirectionEnabled {
+				// Configured for override
+				w.TunnelProtocol = workloadapi.TunnelProtocol_HBONE
+			}
+			// Otherwise supports tunnel directly
+			if model.SupportsTunnel(se.Labels, model.TunnelHTTP) {
+				w.TunnelProtocol = workloadapi.TunnelProtocol_HBONE
+				w.NativeTunnel = true
+			}
+			res = append(res, model.WorkloadInfo{Workload: w, Labels: se.Labels, Source: kind.WorkloadEntry})
+		}
+		return res
+	}, krt.WithName("ServiceEntryWorkloads"))
+	Workloads := krt.JoinCollection([]krt.Collection[model.WorkloadInfo]{PodWorkloads, WorkloadEntryWorkloads, ServiceEntryWorkloads}, krt.WithName("Workloads"))
+	return Workloads
+}
+
+func fetchPeerAuthentications(
+	ctx krt.HandlerContext,
+	PeerAuths krt.Collection[*securityclient.PeerAuthentication],
+	meshCfg *MeshConfig,
+	ns string,
+	matchLabels map[string]string,
+) []*securityclient.PeerAuthentication {
+	return krt.Fetch(ctx, PeerAuths, krt.FilterGeneric(func(a any) bool {
+		pol := a.(*securityclient.PeerAuthentication)
+		if pol.Namespace == meshCfg.GetRootNamespace() && pol.Spec.Selector == nil {
+			return true
+		}
+		if pol.Namespace != ns {
+			return false
+		}
+		sel := pol.Spec.Selector
+		if sel == nil {
+			return true // No selector matches everything
+		}
+		return labels.Instance(sel.MatchLabels).SubsetOf(matchLabels)
+	}))
+}
+
+func fetchWaypoints(ctx krt.HandlerContext, Waypoints krt.Collection[Waypoint], ns, sa string) []Waypoint {
+	return krt.Fetch(ctx, Waypoints,
+		krt.FilterNamespace(ns), krt.FilterGeneric(func(a any) bool {
+			w := a.(Waypoint)
+			return w.ForServiceAccount == "" || w.ForServiceAccount == sa
+		}))
+}
+
+func (a *index) constructServicesFromWorkloadEntry(p *networkingv1alpha3.WorkloadEntry, services []model.ServiceInfo) map[string]*workloadapi.PortList {
+	res := map[string]*workloadapi.PortList{}
+	for _, svc := range services {
+		n := namespacedHostname(svc.Namespace, svc.Hostname)
+		pl := &workloadapi.PortList{}
+		res[n] = pl
+		for _, port := range svc.Ports {
+			targetPort := port.TargetPort
+			if named, f := svc.PortNames[int32(port.ServicePort)]; f {
+				// get port name or target port
+				nv, nf := p.Ports[named.PortName]
```

Comment:
This is not consistent with pod, which only compare with targetPortName if it is not a int type.

Reply:
:+1: